### PR TITLE
[SPARK-24576][BUILD] Upgrade Apache ORC to 1.5.2

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This issue aims to upgrade Apache ORC library from 1.4.4 to 1.5.2 in order to bring the following benefits into Apache Spark.

- [ORC-91](https://issues.apache.org/jira/browse/ORC-91) Support for variable length blocks in HDFS (The current space wasted in ORC to padding is known to be 5%.)
- [ORC-344](https://issues.apache.org/jira/browse/ORC-344) Support for using Decimal64ColumnVector

In addition to that, Apache Hive 3.1 and 3.2 will use ORC 1.5.1 ([HIVE-19669](https://issues.apache.org/jira/browse/HIVE-19465)) and 1.5.2 ([HIVE-19792](https://issues.apache.org/jira/browse/HIVE-19792)) respectively. This will improve the compatibility between Apache Spark and Apache Hive by sharing the common library.

## How was this patch tested?

Pass the Jenkins with all existing tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala [hunk 1]
```diff
@@ -59,6 +59,19 @@ private[sql] object OrcFileFormat {
...
+  def getQuotedSchemaString(dataType: DataType): String = dataType match {
+    case _: AtomicType => dataType.catalogString
+    case StructType(fields) =>
+      fields.map(f => s"`${f.name}`:${getQuotedSchemaString(f.dataType)}")
+        .mkString("struct<", ",", ">")
+    case ArrayType(elementType, _) =>
+      s"array<${getQuotedSchemaString(elementType)}>"
+    case MapType(keyType, valueType, _) =>
+      s"map<${getQuotedSchemaString(keyType)},${getQuotedSchemaString(valueType)}>"
+    case _ => // UDT and others
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala]: nit: Seems the first `_: AtomicType` can be saved because this covers all other cases?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcSerializer.scala [hunk 2]
```diff
@@ -223,6 +223,6 @@ class OrcSerializer(dataSchema: StructType) {
    * Return a Orc value object for the given Spark schema.
    */
   private def createOrcValue(dataType: DataType) = {
-    OrcStruct.createValue(TypeDescription.fromString(dataType.catalogString))
+    OrcStruct.createValue(TypeDescription.fromString(OrcFileFormat.getQuotedSchemaString(dataType)))
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcSerializer.scala]: Why this change?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcSerializer.scala]: Thank you for review, @viirya . ORC 1.5 checks the field name syntax more strictly. For example, a field name having `dot`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala [hunk 3]
```diff
@@ -59,6 +59,19 @@ private[sql] object OrcFileFormat {
...
+      dataType.catalogString
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala]: We don't need to recursively quote `udt.sqlType`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcSerializer.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcSerializer.scala]: @dongjoon-hyun Thanks for explaining it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala]: Thank you for review, @maropu . Yes, that is not handled here because the goal is to support user's column names like `col1.x` usually at the top-level column names.

**Comment by dongjoon-hyun** [comment]
@dbtsai . This seems to be another difference due to recent build system changes.
- build/mvn -Phive clean package -DskipTests (Build Success)
- build/sbt -Phive clean package (Build Failure)

I'll take a look at this.

**Comment by viirya** [comment]
@dongjoon-hyun What error you see, I can run the build with sbt without problem.

**Comment by dongjoon-hyun** [comment]
It occurs when we uses `classifier`. This PR uses `nohive` classifier for `orc-core`. If you try the above commands on this PR, the build fails for `sbt` only.

File: sql/core/pom.xml [hunk 4]
```diff
@@ -90,11 +90,39 @@
...
+          <artifactId>hadoop-hdfs</artifactId>
+        </exclusion>
+        <!--
+          orc-core:nohive doesn't have this dependency, but we adds this to prevent
+          sbt from getting confused.
+        -->
+        <exclusion>
+          <groupId>org.apache.hive</groupId>
+          <artifactId>hive-storage-api</artifactId>
+        </exclusion>
```
Comment (dongjoon-hyun) [inline sql/core/pom.xml]: I added the above eight lines to be consistent for both `mvn` and `sbt`.

**Comment by dongjoon-hyun** [comment]
@gatorsmile . For benchmark, ORC community is officially working on [ORC-386](https://github.com/apache/orc/pull/290). After it's finalized, I'll try to share you based on that.

**Comment by dongjoon-hyun** [comment]
@gatorsmile .  [ORC-386](https://github.com/apache/orc/pull/290) is not merged yet, but I've got the following from Hive's `DecimalBench`. For Spark side, we can proceed after this BUILD PR.
```
Benchmark               (version)  Mode  Cnt         Score       Error  Units
DecimalBench.read        ORIGINAL  avgt    5  10375991.566 ± 99751.161  us/op
DecimalBench.read   USE_DECIMAL64  avgt    5   7368917.561 ± 39257.332  us/op
DecimalBench.write       ORIGINAL  avgt    5    211008.673 ± 30220.403  us/op
DecimalBench.write  USE_DECIMAL64  avgt    5     33881.693 ±   216.601  us/op
```

**Comment by gatorsmile** [comment]
@dongjoon-hyun Could you give how the benchmark works? What is the workload pattern? How does the benchmark invoke Spark?

**Comment by gatorsmile** [comment]
@dongjoon-hyun Could you submit a PR to use the latest Decimal64ColumnVector like what Hive does https://issues.apache.org/jira/browse/HIVE-19629 ?

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @gatorsmile . I will proceed.
Also, thank you, @viirya and @maropu .
